homeassistant/components/cloud/backup.py

```diff
@@ -0,0 +1,166 @@
+"""Backup platform for the cloud integration."""
+
+from __future__ import annotations
+
+import base64
+from collections.abc import AsyncIterator, Callable, Coroutine
+import hashlib
+from typing import Any
+
+from aiohttp import ClientResponseError
+from hass_nabucasa import Cloud
+from hass_nabucasa.cloud_api import (
+    async_files_delete_file,
+    async_files_download_details,
+    async_files_list,
+    async_files_upload_details,
+)
+
+from homeassistant.components.backup import AgentBackup, BackupAgent, BackupAgentError
+from homeassistant.core import HomeAssistant, callback
+
+from .client import CloudClient
+from .const import DATA_CLOUD, DOMAIN
+
+_STORAGE_BACKUP = "backup"
+
+
+async def _b64md5(stream: AsyncIterator[bytes]) -> str:
+    """Calculate the MD5 hash of a file."""
+    file_hash = hashlib.md5()
+    async for chunk in stream:
+        file_hash.update(chunk)
+    return base64.b64encode(file_hash.digest()).decode()
+
+
+async def async_get_backup_agents(
+    hass: HomeAssistant,
+    **kwargs: Any,
+) -> list[BackupAgent]:
+    """Return the cloud backup agent."""
+    return [CloudBackupAgent(hass=hass, cloud=hass.data[DATA_CLOUD])]
+
+
+class CloudBackupAgent(BackupAgent):
+    """Cloud backup agent."""
+
+    name = DOMAIN
+
+    def __init__(self, hass: HomeAssistant, cloud: Cloud[CloudClient]) -> None:
+        """Initialize the cloud backup sync agent."""
+        super().__init__()
+        self._cloud = cloud
+        self._hass = hass
+
+    @callback
+    def _get_backup_filename(self) -> str:
+        """Return the backup filename."""
+        return f"{self._cloud.client.prefs.instance_id}.tar"
+
+    async def async_download_backup(
+        self,
+        backup_id: str,
+        **kwargs: Any,
+    ) -> AsyncIterator[bytes]:
+        """Download a backup file.
+
+        :param backup_id: The ID of the backup that was returned in async_list_backups.
+        :param path: The full file path to download the backup to.
+        """
+        if not self._cloud.is_logged_in:
+            raise BackupAgentError("Not logged in to cloud")
+
+        if not await self.async_get_backup(backup_id):
+            raise BackupAgentError("Backup not found")
+
+        details = await async_files_download_details(
```

Comment:
We need to handle aiohttp client exceptions for all the library calls.

Reply:
Fixed in ce383e8986af9188e6ddb0840e30ab557ad29c61